src/vm/types/mod.rs

```diff
@@ -0,0 +1,449 @@
+        if buff_data.len() > u32::max_value() as usize {
+            Err(RuntimeErrorType::BufferTooLarge.into())
+        } else if buff_data.len() as i128 > MAX_VALUE_SIZE {
+            Err(RuntimeErrorType::ValueTooLarge.into())
+        } else {
+            Ok(Value::Buffer(BuffData { data: buff_data }))
+        }
+    }
+
+    pub fn size(&self) -> Result<i128> {
+        match self {
+            Value::Int(_i) => AtomTypeIdentifier::IntType.size(),
+            Value::Bool(_i) => AtomTypeIdentifier::BoolType.size(),
+            Value::Principal(_) => AtomTypeIdentifier::PrincipalType.size(),
+            Value::Buffer(ref buff_data) => Ok(buff_data.data.len() as i128),
+            Value::Tuple(ref tuple_data) => tuple_data.size(),
+            Value::List(ref list_data) => list_data.type_signature.size(),
+            Value::Optional(ref opt_data) => opt_data.type_signature().size(),
+            Value::Response(ref res_data) => res_data.type_signature().size()
+        }
+    }
+
+}
+
+impl fmt::Display for OptionalData {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self.data {
+            Some(ref x) => write!(f, "(some {})", x),
+            None => write!(f, "none")
+        }
+    }
+}
+
+impl fmt::Display for ResponseData {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self.committed {
+            true => write!(f, "(ok {})", self.data),
+            false => write!(f, "(err {})", self.data)
+        }
+    }
+}
+
+impl fmt::Display for Value {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Value::Int(int) => write!(f, "{}", int),
+            Value::Bool(boolean) => write!(f, "{}", boolean),
+            Value::Buffer(vec_bytes) => write!(f, "0x{}", hash::to_hex(&vec_bytes.data)),
+            Value::Tuple(data) => write!(f, "{}", data),
+            Value::Principal(principal_data) => write!(f, "{}", principal_data),
+            Value::Optional(opt_data) => write!(f, "{}", opt_data),
+            Value::Response(res_data) => write!(f, "{}", res_data),
+            Value::List(list_data) => {
+                write!(f, "(")?;
+                for (ix, v) in list_data.data.iter().enumerate() {
+                    if ix > 0 {
+                        write!(f, " ")?;
+                    }
+                    write!(f, "{}", v)?;
+                }
+                write!(f, ")")
+            }
+        }
+    }
+}
+
+impl PrincipalData {
+    pub fn parse_qualified_contract_principal(literal: &str) -> Result<PrincipalData> {
+        let split: Vec<_> = literal.splitn(2, ".").collect();
+        if split.len() != 2 {
+            return Err(RuntimeErrorType::ParseError(
+                "Invalid principal literal: expected a `.` in a qualified contract name".to_string()).into());
+        }
+        let sender = Self::parse_standard_principal(split[0])?;
+        let name = split[1].to_string();
```

Comment:
Can we be sure that `name` is an ASCII string?

Reply:
Not as written -- but that's going to be addressed in #1078 (is my understanding), at which point we should have a simple guarded string type.